36 [Code][Gameplay] Cannon

DrakeHunterK7 opened the pull request:
- Created IInteractable Interface for use by all Interactables
- Created Slingshot/Cannon using the aforementioned interface. 
- Created Cannonball prefab for Slingshot


Instructions:
Press 'E' to begin using the slingshot when standing close to the Slingshot. Left Click to start aiming the Slingshot then let go of Left Click to shoot. Press 'E' while using it to forfeit control of the Slingshot.

George540 commented on Project Bunny/Assets/Scripts/Player/Cannon.cs:
```diff
@@ -0,0 +1,255 @@
+        private GameObject _currentCannonBall;
+        private bool _isActive;
+        private GameObject _player;
+        private StudentController _playerScript;
```
name it _studentController, so it's easy to access it without having to think how it's named

George540 commented on Project Bunny/Assets/Scripts/Player/Cannon.cs:
```diff
@@ -0,0 +1,255 @@
+        private GameObject _player;
+        private StudentController _playerScript;
+        private Camera _playerCam;
+        private CinemachineVirtualCamera _cameraSettings;
```
Same here

George540 commented on Project Bunny/Assets/Scripts/Player/Cannon.cs:
```diff
@@ -0,0 +1,255 @@
+        {
+            //Initialize key variables
+            _isActive = true;
+            _player = GameObject.FindWithTag("Player");
```
Avoid using Find() Can you make a function that finds the reference from the StudentController? I think on the OnInteract function, you can pass the StudentController.cs component as a parameter

George540 commented on Project Bunny/Assets/Scripts/Player/Cannon.cs:
```diff
@@ -0,0 +1,255 @@
+            //Initialize key variables
+            _isActive = true;
+            _player = GameObject.FindWithTag("Player");
+            _playerScript = _player.GetComponent<StudentController>();
```
As the comment mentioned above, same here

George540 commented on Project Bunny/Assets/Scripts/Player/Cannon.cs:
```diff
@@ -0,0 +1,255 @@
+            }
+            
+            //Disable player controller in order to set the player's position manually
+            _playerCharController = _player.GetComponent<CharacterController>();
```
Same here

George540 commented on Project Bunny/Assets/Scripts/Player/Cannon.cs:
```diff
@@ -0,0 +1,255 @@
+            _playerScript = null;
+            
+            //Get the Body component of the player's CinemachineVirtualCamera.
+            var componentBase = _cameraSettings.GetCinemachineComponent(CinemachineCore.Stage.Body);
```
Can you make a camera script that accesses these components easily instead of having to do GetCinemachineComponent all the time?

George540 commented on Project Bunny/Assets/Scripts/Player/Cannon.cs:
```diff
@@ -0,0 +1,255 @@
+        /// <summary>
+        /// Throw the snowball and restore related variables to default values.
+        /// </summary>
+        private void ThrowSnowball()
```
A lot of the code is from the StudentController.cs script. Why not use those instead of copy-pasting them here?

George540 commented on Project Bunny/Assets/Scripts/Player/Cannon.cs:
```diff
@@ -0,0 +1,255 @@
+        /// </summary>
+        /// <param name="mousePos"></param>
+        /// <returns></returns>
+        private float MousePosToRotationInput(Vector2 mousePos)
```
Same for this one. You can reuse them

George540 commented on Project Bunny/Assets/Scripts/Player/StudentController.cs:
```diff
@@ -297,6 +342,29 @@ private float MousePosToRotationInput(Vector2 mousePos)
+        /// Utility function that returns the nearest interactable to the player
+        /// </summary>
+        /// <returns></returns>
+        private IInteractable ReturnNearestInteractable() 
```
Why not use the OnCharacterCollisionHit() function callback instead? You can handle multiple interactables at once, but usually none of them will be close to one another

DrakeHunterK7 replied:
The idea here is to call the ReturnNearestInteractable() function when the player pressesn the 'E' key, and not when the player collides with the gadget. If that makes sense.

George540 reviewed (changes requested):
Good work! But there are some redundancies that can be avoided. Will make the transfer to the network easier

Cappic-01 reviewed (commented):
Looks good to me excluding the requested changes by George. Will rereview and approve once a review will be requested again.

DrakeHunterK7 commented:
I removed some redundancies but I couldn't simplify the Cannonball throwing and aiming logic code since even though the code is doing the same things as StudentController.cs, the variables affected are different for both the StudentController.cs and Cannon.cs so it's safer to just keep the Cannonball throwing and aiming logic the way it is. 

Also added a Utilities script to add all utilities that can be used by multiple scripts.

George540 reviewed (approved):
Good work!

KristianKurk commented on Project Bunny/Assets/Scripts/Player/Cannon.cs:
```diff
@@ -0,0 +1,228 @@
+        private StudentController _currentStudentController;
+        private CinemachineComponentBase _playerVCamSettings;
+        private CharacterController _playerCharController;
+        [SerializeField] [Min(0)] private float _rotationSpeed;
```
group your serializefield variables if you can

KristianKurk reviewed (approved):
looks clean, left a little comment but it's just an organizational thing
